packages/ui/src/hooks/use-social.ts

```diff
@@ -0,0 +1,184 @@
+  const callbackUriScheme = logtoNativeSdk?.callbackUriScheme;
+
+  const state: State = { uunid, platform, callbackUriScheme };
+
+  return Buffer.from(JSON.stringify(state)).toString('base64');
+};
+
+export const decodeState = (state: string) => {
+  try {
+    const value = JSON.parse(Buffer.from(state, 'base64').toString()) as State;
+
+    return value;
+  } catch {}
+};
+
+export const stateValidation = (state: string, connectorId: string) => {
+  const stateStorage = sessionStorage.getItem(`${storageKeyPrefix}:${connectorId}`);
+
+  return Boolean(stateStorage) && stateStorage === state;
+};
+
+export const storeState = (state: string, connectorId: string) => {
+  sessionStorage.setItem(`${storageKeyPrefix}:${connectorId}`, state);
+};
+
+/* ============================================================================ */
+
+const isNativeWebview = () => {
+  return ['ios', 'android'].includes(logtoNativeSdk?.platform ?? '');
+};
+
+const useSocial = () => {
+  const { setToast } = useContext(PageContext);
+  const parameters = useParams();
+
+  const { result: invokeSocialSignInResult, run: asyncInvokeSocialSignIn } =
+    useApi(invokeSocialSignIn);
+
+  const { result: signInWithSocialResult, run: asyncSignInWithSocial } = useApi(signInWithSocial);
+
+  const invokeSocialSignInHandler = useCallback(
+    async (connectorId: string) => {
+      const state = generateState();
+      storeState(state, connectorId);
+
+      const { origin } = window.location;
+
+      return asyncInvokeSocialSignIn(connectorId, state, `${origin}/callback/${connectorId}`);
+    },
+    [asyncInvokeSocialSignIn]
+  );
+
+  const signInWithSocialHandler = useCallback(
+    (connectorId: string, state: string, code: string) => {
+      if (!stateValidation(state, connectorId)) {
+        // TODO: Invalid state error message
+        return;
+      }
+      void asyncSignInWithSocial({ connectorId, state, code, redirectUri: '' });
+    },
+    [asyncSignInWithSocial]
+  );
+
+  const socialCallbackHandler = useCallback(
+    (connectorId?: string) => {
+      const { state, code, error, error_description } = parseQueryParameters(
+        window.location.search
+      );
+
+      if (error) {
+        setToast(`${error}${error_description ? `: ${error_description}` : ''}`);
+      }
+
+      if (!state || !code || !connectorId) {
+        // TODO: error message
+        return;
+      }
+
+      const decodedState = decodeState(state);
+
+      if (!decodedState) {
+        // TODO: invalid state error message
+        return;
+      }
+
+      const { platform, callbackUriScheme } = decodedState;
+
+      if (platform === 'web') {
+        signInWithSocialHandler(connectorId, state, code);
+
+        return;
+      }
+
+      if (!callbackUriScheme) {
+        // TODO: native callbackUriScheme not found error message
+        return;
+      }
+
+      window.location.assign(`${callbackUriScheme}${connectorId}${window.location.search}`);
```

Comment:
per offline discussion, maybe hard code it `logto-callback://${appId}/web?${search}`.
https://silverhand-io.slack.com/archives/C02A8G4HVAM/p1650343181926779

Reply:
updated